File: Question5.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

bool twosum(const vector<int>& arr, int target) {
    // Create a hash map to store the elements and their indices. Hashmap is similar to dictionaries in py
    unordered_map<int, int> elements;

    // Iterate through the array
    for (int i = 0; i < arr.size(); ++i) {
        int complementno = target - arr[i];

        // Check if the complement exists in the hash map
        if (elements.find(complementno) != elements.end()) {
            // Print the indices of the elements that add up to the target
            cout << "Pair found at indices " << elements[complementno] << " and " << i << endl;
            return true;
        }

        // Add the current element to the hash map
        elements[arr[i]] = i;
    }

    // No pair found
    cout << "No pair found" << endl;
    return false;
}

int main() {
    vector<int> arr = {17, 12, 1, 45};
    int target = 46;

    twosum(arr, target);

    return 0;
}

File: Question5.py
def twosum(nums, target):
    # Creating a dictionary to store the value and its index
    num_index = {}

    # Iterate over the array
    for i, num in enumerate(nums):
        # Calculate the complement of the current number
        complement_no = target - num

        # Check if the complement is already in the dictionary
        if complement_no in num_index:
            # If found, return the indices
            return [num_index[complement_no], i], [complement_no, num]

        # Otherwise, add the current number and its index to the dictionary
        num_index[num] = i

    # If no solution is found, return an empty list
    return []


nums = [5, 8, 15, 29, 16]
target = 25
result = twosum(nums, target)
print(result)
